sanitized_dump/utils/db.py

```diff
@@ -1,41 +1,93 @@
 
-    if not default_db:
-        raise ValueError("No default database found")
+def read_mysql_options_from_path(path):
+    options = SafeConfigParser()
+    options.read(path)
+    keys = ["user", "password", "host", "port", "database"]
+    result = {}
+    for key in keys:
+        if options.has_option("client", key):
+            result[key] = options.get("client", key)
+    return result
 
-    engine = default_db.get('ENGINE')
-    db_name = default_db.get('NAME')
-    db_user = default_db.get('USER')
-    db_password = default_db.get('PASSWORD')
-    db_host = default_db.get('HOST')
-    db_port = default_db.get('PORT')
+class DatabaseUrlBuilder(object):
 
-    if not engine or not db_name:
-        raise ValueError("Database configuration not supported")
+    def __init__(self, engine, user, password, host, port, database):
+        self.engine = engine
+        self.user = user
+        self.password = password
+        self.host = host
+        self.port = port
+        self.database = database
 
-    database_type = None
-    for db_type in DATABASE_TYPES:
-        if db_type in engine:
-            database_type = db_type
-            break
+    def to_string(self):
+        if not all([self.engine, self.database, self.host, self.port]):
+            raise ValueError("Database configuration not supported")
+
+        has_login = (self.user or self.password)
+
+        login_part = '{user}:{password}@' if has_login else ''
+        port_part = ':{port}' if self.port else ''
+
+        database_string_template = (
+            '{engine}://' + login_part + '{host}' + port_part + '/{db}')
 
-    if not database_type:
-        raise ValueError('Database type "{}" is not supported'.format(engine))
-
-    login_part = '{username}:{password}@' if (db_user or db_password) else ''
-    port_part = ':{port}' if db_port else ''
-    database_string_template = (
-        '{type}://' + login_part + '{host}' + port_part + '/{db}')
-
-    return database_string_template.format(
-        type=database_type,
-        username=db_user,
-        password=db_password,
-        host=db_host,
-        port=db_port,
-        db=db_name,
-    )
+        return database_string_template.format(
+            engine=self.engine,
+            user=self.user,
+            password=self.password,
+            host=self.host,
+            port=self.port,
+            db=self.database,
+        )
+
+    @classmethod
+    def create_from_django_database(cls, django_database):
+        engine = get_database_engine_from_django_database(django_database)
+        if not engine:
+            raise ValueError("Database configuration not supported")
+
+        kwargs = {
+            "engine": engine,
+            "user": django_database.get("USER"),
+            "password": django_database.get("PASSWORD"),
+            "host": django_database.get("HOST", None) or "localhost",
+            "port": django_database.get("PORT", None) or DEFAULT_PORTS.get(engine),
```

Comment:
Why do you need DEFAULT_PORTS? IMHO if the settings doesn't specify a port, then the database URL shouldn't specify a port either.

Reply:
Fixed